core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddLocalExchanges.java

```diff
@@ -534,32 +535,32 @@ public PlanWithProperties visitTopNRanking(TopNRankingNode node, StreamPreferred
         @Override
         public PlanWithProperties visitTableWriter(TableWriterNode node, StreamPreferredProperties parentPreferences)
         {
-            StreamPreferredProperties requiredProperties;
-            StreamPreferredProperties preferredProperties;
-            // TODO: add support for arbitrary partitioning in local exchanges
-            if (getTaskWriterCount(session) > 1) {
-                boolean hasFixedHashDistribution = node.getPartitioningScheme()
-                        .map(scheme -> scheme.getPartitioning().getHandle())
-                        .filter(isEqual(FIXED_HASH_DISTRIBUTION))
-                        .isPresent();
-                if (node.getPartitioningScheme().isEmpty()) {
-                    requiredProperties = fixedParallelism();
-                    preferredProperties = fixedParallelism();
-                }
-                else if (hasFixedHashDistribution) {
-                    requiredProperties = exactlyPartitionedOn(node.getPartitioningScheme().get().getPartitioning().getColumns());
-                    preferredProperties = requiredProperties;
-                }
-                else {
-                    requiredProperties = singleStream();
-                    preferredProperties = defaultParallelism(session);
-                }
+            if (getTaskWriterCount(session) == 1) {
+                return planAndEnforceChildren(node, singleStream(), defaultParallelism(session));
```

Comment:
Extract preparatory commit flipping the `if (getTaskWriterCount(session) > 1) {` condition, otherwise it's unclear what is being changed here.